### PR TITLE
Request response

### PR DESCRIPTION
**Pull request by Aoxian**
Added the concept of a Request and a Response.  Also made the Parser return a Request.  Next step is to flush out a Router and Controller (and get rid of the Handler) in a separate PR.

### CODE REVIEW

File: src/Aoxian.Web.HTTP/RequestParser.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Aoxian.Web.HTTP
+{
+    public class RequestParser : IRequestParser
+    {
+        private Request _request;
```
Comment (td-extreme) [inline src/Aoxian.Web.HTTP/RequestParser.cs]: class variables should be used to hold state between method calls by the user. Your only public method resets this value right after it is called. 
Get rid of this line.

File: src/Aoxian.Web.HTTP/RequestParser.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        public Request Parse(byte[] requestBuffer)
+        {
+            _request = new Request();
```
Comment (td-extreme) [inline src/Aoxian.Web.HTTP/RequestParser.cs]: You should look into using the object initializer.
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer

File: src/Aoxian.Web.HTTP/RequestParser.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            var requestString = CleanRequest(
+                DecodeRequestBuffer(requestBuffer)
+            );
+            SplitRequestAndBody(requestString, out var requestLines, out var body);
+            ParseLines(requestLines);
+            ParseBody(body);
+            return _request;
+        }
+
+        private string DecodeRequestBuffer(byte[] request)
```
Comment (td-extreme) [inline src/Aoxian.Web.HTTP/RequestParser.cs]: `private string DecodeRequestBuffer(byte[] requestBuffer)`

Try and be more specific with parameter names. 
Here `request` is type of `byte[]`
Line 30 `request` is type of `string`
Also you have a global `_request` that is type of `Request`

File: src/Aoxian.Web.HTTP/RequestParser.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        {
+            return Encoding.UTF8.GetString(request, 0, request.Length);
+        }
+
+        private static string CleanRequest(string request)
+        {
+            return request.Replace("\r\n", "\n");
+        }
+
+        private static void SplitRequestAndBody(string cleanedRequest, out List<string> lines, out string body)
```
[the remark this answers is not in the document]
Comment (Aoxian) [reply src/Aoxian.Web.HTTP/RequestParser.cs]: I wanted to do a `{lines, body} = SplitRequestAndBody(TheEntireRequest)` kind of thing.  The `out` variables allowed me to return two things.  I was not aware of tuples in C#; reading up now.  It seems they are different than tuples like I am used to them.

File: src/Aoxian.Web.HTTP/RequestParser.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        private const int PATH_INDEX = 1;
+        private const char REQUEST_PARAMATER_SEPERATOR = ' ';
+        private const string HEADER_SEPERATOR = ": ";
+        private const char NEW_LINE = '\n';
+        private const string CHARAGE_RETURN = "\r";
+        private const string EMPTY_LINE = "\n\n";
+
+        public Request Parse(byte[] requestBuffer)
+        {
+            var rawRequest = Encoding.UTF8.GetString(requestBuffer, 0, requestBuffer.Length)
```
Comment (td-extreme) [inline src/Aoxian.Web.HTTP/RequestParser.cs]: I feel like this is an SRP violation. Move this somewhere else and change this method to take the `rawRequest` as a `string` with `\r` already removed.

 `public Request Parse(string rawRequest)`

**Pull request merged**